xdsl/dialects/func.py

```diff
@@ -47,6 +51,64 @@ def verify_(self) -> None:
                 "input types"
             )
 
+    @classmethod
+    def parse(cls, parser: Parser) -> FuncOp:
+        # Parse visibility keyword if present
+        visibility = parser.tokenizer.next_token_of_pattern(
+            re.compile("public|nested|private")
+        )
+        if isinstance(visibility, Span):
+            visibility = visibility.text
+        else:
+            visibility = "private"
+
+        # Parse function name
+        name = parser.parse_symbol_name().data
+
+        # Parse function arguments
+        args = parser.parse_comma_separated_list(
+            parser.Delimiter.PAREN,
+            lambda: parser.parse_optional_argument() or parser.parse_type(),
```

Comment:
For better errors, we could have something like `parser.parse_optional_argument() or parser.parse_optional_type() or parser.raise_error("Expected argument or type")`

Reply:
Except parse_optional_type is not a thing :( I can go with try_parse_type, but am not sure about the lexer/tokenizer status there.
I'll just use try_parse_type meanwhile unless it's an issue!